Paper: Solving Polynomial Systems with phcpy

janverschelde opened the pull request:
With this PR we submit a draft of our paper
Solving Polynomial Systems with phcpy

Sincerely Yours,
 Jan Verschelde

janverschelde committed:
a very modest start of the paper ...

janverschelde committed:
extended reference list, added two paragraphs of text and a figure

janverschelde committed:
added section on CGI scripting and more references

janverschelde committed:
changed python into python3

janverschelde committed:
added one section on jupyterhub and more references

janverschelde committed:
added a figure and a section on mechanism design

janverschelde committed:
added a section on mission and realization

janverschelde committed:
added the beginning of a section on quality up

janverschelde committed:
added references to MixedVol, DEMiCs, QDlib

janverschelde committed:
added version numbers, jump from 0.1.5 to 0.9.4

janverschelde committed:
added another beginning of a section to the skeleton of the paper

JazzTap committed:
nest some sections

JazzTap committed:
gathering past and likely future uses of the library, by literature search.

janverschelde committed:
Merge pull request #1 from JazzTap/2019

Three-act structure & set up lit search

janverschelde committed:
Eliminated the Belief Propagation section because the calculations in that
paper are perform with HOM4PS-3, which is closed source software.
Started a section on Related Software, will add references to the
literature on those three software packages later.

janverschelde committed:
added more references in the related software section

janverschelde committed:
expanded the quality up section

JazzTap committed:
tuesday kickoff

JazzTap committed:
citation pass

JazzTap committed:
Merge branch '2019' of https://github.com/JazzTap/scipy_proceedings into 2019

janverschelde committed:
renamed the section title Intensive Methods into Solving Polynomial Systems,
added a subsection on polynomial homotopy continuation

JazzTap committed:
reference wrangling

JazzTap committed:
Merge remote-tracking branch 'verschelde/2019' into 2019

JazzTap committed:
incorporate survey of numerical alg geom

janverschelde committed:
Merge pull request #2 from JazzTap/2019

Add to intro for general audience

janverschelde committed:
updated version number to 0.9.5 and number of user accounts to 146

JazzTap committed:
bibliography - applications

JazzTap committed:
Merge remote-tracking branch 'verschelde/2019' into 2019

JazzTap committed:
direct manipulation, systems biology

JazzTap committed:
direct manipulation figure

JazzTap committed:
continue staging references

JazzTap committed:
rigid graph theory

janverschelde committed:
added the use of the cyclic 4-roots problem in the SymPy online documentation

JazzTap committed:
Merge remote-tracking branch 'verschelde/2019' into 2019

janverschelde committed:
mentioned CASC 2018 paper after solving the cyclic 8-roots problem

janverschelde committed:
new picture visualizes elapsed performances

janverschelde committed:
inserted figure on elapsed performances

JazzTap committed:
more applications

JazzTap committed:
close out examples

JazzTap committed:
Merge branch '2019' into 2019

janverschelde committed:
Merge pull request #3 from JazzTap/2019

Finish drafting examples

janverschelde committed:
made a first pass through the paper, editing the mission section,
removing the HOM4PS-3 (because closed source) from related software,
alphabetized the list of references, added a reference to deflation,
added semi colons in the enumerate sections

janverschelde committed:
added an explicit reference in the text to the Bertini book

loldja committed:
Initial Review submission

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+this paper is based on version 0.9.5 of phcpy,
+whereas version 0.1.5 was current at the time of [Ver14]_.
+An example of these changes is that the software described in [SVW03]_ 
+was recently parallelized for phcy [Ver18]_.
```
s/phcy/phcpy/

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+The mission of phcpy is to bring polynomial homotopy continuation
+into Python's computational ecosystem.
+
+phcpy wraps the shared object files of a compiled PHCpack, 
```
It looks very odd to have a lowercase first letter in a paragrph. Most style guides would say capilitalize even if it a usually lowecase name. I am not dogmatic about this, so do what you feel is right.

deniederhut replied:
Geez you wrote the interfaces by hand. This isn't crucial for the paper, but I am curious why you didn't use something like swig or cython.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+----------------
+
+PHCpack is one of three FOSS packages for polynomial homotopy computation currently under development. Of these, only Bertini 2 [Bertini2.0]_ also offers Python bindings, although it is not GPU-accelerated and does not export the numerical irreducible decomposition, among other differences.
+Version 1.4 of Bertini (released under a too restrictive license)
```
Leave out the editorializing of "too restrictive." That is a subjective judgment dependent on the desired purpose. At the most you could say "more restrictive." But it adds nothing as the current version of Bertini is GPL3. Some dependencies of Bertini 2 are more restrictive than GPL3, but then the argument gets complicated. I would advocate eliminating the parenthetical phrase entirely as it's not germane to your topic.

JazzTap replied:
Has been dropped

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+
+In our first design of a web interface to ``phc``, we developed a collection of Python scripts (mediated through HTML forms), following common programming patterns [Chu06]_.
+
+MySLQdb does the management of user data, including a) names and encrypted passwords, b) generic, random folder names to store data files, and c) file names with polynomial systems they have solved. With the module smtplib, we defined email exchanges for an automatic 2-step registration process and password recovery protocol.
```
Maybe provide references for MySQLdb. At first I thought is was a misspelling of MySQL db as I had not previously heard of this package. smtplib is standard library, so it's not as important to have a reference for that, although it doesn't hurt if you decide to.

JazzTap replied:
Package references added throughout, except smtplib.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+
+MySLQdb does the management of user data, including a) names and encrypted passwords, b) generic, random folder names to store data files, and c) file names with polynomial systems they have solved. With the module smtplib, we defined email exchanges for an automatic 2-step registration process and password recovery protocol.
+
+As of the middle of May 2019, our web server has 146 user accounts, each having access to our new JupyterHub instance.
```
JupyterHub itself may also bear a reference, even though you have a reference for how you use JupyterHub with phcpy.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+
+With JupyterHub, we provide online access [Pascal]_ to environments with Python and SageMath kernels pre-installed, both featuring phcpy and tutorials on its use (per next section). Since Jupyter is language-agnostic, execution environments in several dozen languages are possible. Our users can also run code in a Python Terminal session.
+
+For the user administration, we refreshed our first web interface. A custom JupyterHub Authenticator connects to the existing MySQL database, and triggers a SystemdSpawner that isolates the actions of users to separate processes and logins in generic home folders.
```
Lose the comma after "database."

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+
+For the user administration, we refreshed our first web interface. A custom JupyterHub Authenticator connects to the existing MySQL database, and triggers a SystemdSpawner that isolates the actions of users to separate processes and logins in generic home folders.
+
+The account management prompts by e-mail were hooked to new Tornado Handler instances, which perform account registration and activation on the database, as well as password recovery and reset. Each pathway serves HTML seamlessly with the JupyterHub interface, with their forms living in new extensions of its Jinja templates.
```
s/Handler/RequestHandler/
Also, "pathway?" That's not a Jupyter or Tornado term as I know it. Reword or clarify.
Also, "their" and "its" refer to what? Reword or clarify.

JazzTap replied:
Applied. Reworded to 'route'. Addressed the 'HTML forms' directly.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+   sols = solve(f)
+   for sol in sols: print(sol)
+
+The first solution of the given trinomial can be read off as (0.48613… + 0.0i, 0.34258… - 0.0i), where the imaginary part of x_0 is exactly zero, and that of y_0 negligibly small. Programmatically, these can be accessed using either ``solve(f, dictionary_output=True)``, or equivalently by parsing strings through ``[phcpy.solutions.strsol2dict(sol) for sol in solve(f)]``.
```
s/be read off as/be read as/

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+Direct Manipulation
+-------------------
+
+One consequence of the Jupyter notebook's rich output is the possibility of rich input, as explored through ipywidgets and interactive plotting libraries. The combination of rich input with fast numerical methods makes surprising interactions possible, such as interactive solution of Apollonius' Problem, which is to construct all circles tangent to three given circles in a plane.
```
Possobly provide a reference for ipywidgets.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+
+The tutorial given in the phcpy documentation was adapted for a demo accompanying a SciPy poster in 2017, whose code [APP]_ will run on our JupyterHub (by copying over ``apollonius_d3.ipynb`` and ``apollonius_d3.js``).
+
+This system of 3 nonlinear constraints in 5 parameters for each of 8 possible tangent circles (some of which have imaginary position or radius in certain configurations), which we solved interactively (Fig. :ref:`apollonius`). In fact, Jupyter is a suitable environment for mapping algebraic inputs to their geometric representations (in a 2D plane), through its interaction with D3.js [D3]_ for non ...
```
This first sentence is incomplete. Essentially it says, "This system, which we solved." The sentences needs a verb and possibly a subject.

JazzTap replied:
Intervening parentheses moved outside of problematic sentence. Revised paragraph.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
s/system on input/system as input/

cbcunc commented on papers/jan_verschelde/phcpy.rst:
s/associated to/associated with/

cbcunc commented on papers/jan_verschelde/phcpy.rst:
All three of these classes of polynomial systems need references to substantiate the claims made about them.

JazzTap replied:
Added by @janverschelde.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
Reference for SymPy or SymPy documentation.

deniederhut commented:
Hi @janverschelde ! We had fewer reviewers complete a review for your paper than we were expecting, so I'll be stepping in. Please expect comments from me some time in the next handful of days 🙂.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
Is it possible to get the code block to appear before the two tables? I don't see anything in your RST to indicate that it shouldn't. But that's not how it's being rendered.

janverschelde replied:
Yes, the insertion of a "..raw:: latex" did the trick.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
"allows to estimate" is unclear. How about simply "The interpolation estimates running times..." or something like that.

JazzTap replied:
s/allows to/allows us to/

cbcunc commented on papers/jan_verschelde/phcpy.rst:
s/One/one/

deniederhut commented on make_paper.sh:
```diff
@@ -7,7 +7,7 @@ if [[ ! -d $DIR ]]; then
 fi
 
-python publisher/build_paper.py $DIR
+python3 publisher/build_paper.py $DIR
```
Please don't modify files outside of your paper's directory

deniederhut commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+This also makes PHCpack accessible from Jupyter notebooks, 
+including a JupyterHub available online [Pascal]_.
+
+The meaning of *solving* evolved from computing approximations to
```
Some commas might might this a bit easier to parse, e.g. "from computing approximations, to all isolated solutions, to the numerical"

deniederhut commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+The meaning of *solving* evolved from computing approximations to
+all isolated solutions into the numerical irreducible decomposition
+of the solution set.  The numerical irreducible decomposition includes
+not only the isolated solutions, but also representations for all
```
"also the representations"

deniederhut commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+The focus of this paper is on the application of new technology
+to solve polynomial systems, in particular, cloud computing [BSVY15]_
+and multicore shared memory parallelism
+accelerated with graphics processing units [VY15]_.
```
Ooh nice 😄

deniederhut commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+Although phcpy has been released for only five years,
+three instances in the research literature of symbolic computation, 
+geometry and topology, and chemical engineering (respectively) 
+mention its application to their computations.
```
:tada:

deniederhut commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+Mission
+---------
+
+The mission of phcpy is to bring polynomial homotopy continuation
```
It might be a good idea to provide a short explanation of homotopy continuation and how it works, for readers who are not familiar.

janverschelde replied:
Added a second paragraph in the introduction to explain homotopy continuation very briefly.

deniederhut commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+Code Snippets
+-------------
+
+Learning a new API is daunting enough without also being a crash course in algebraic geometry. Therefore, the user's manual of phcpy [PHCPY]_ begins with a tutorial section using only the blackbox solver ``phcpy.solver.solve(system, ...)``. In this API, ``system`` is a list of strings representing polynomials, terminated by semicolons, and containing as many variables as equations.
```
Yeah I took a quick trip through your documentation and it is quite extensive 👍

deniederhut commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+Solving Polynomial Systems
+==========================
+
+Our input is a list of polynomials in several variables.
```
Is there a reason you aren't using numpy for passing numeric data back and forth?

janverschelde replied:
The symbolic format for multivariate polynomials is more common and the parsing of the strings is done directly by the code in PHCpack.

deniederhut commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+--------------------------------
+
+As we compute over the field of complex numbers,
+the continuity of the solution set in function of changing
```
I'm having a hard time parsing this sentence -- are we missing an article somewhere?

deniederhut commented on papers/jan_verschelde/phcpy.rst:
Am I reading this correctly that the double-double task takes just as long on 8 cores as it does on a single core?

janverschelde replied:
Thanks for catching that error!

deniederhut commented on papers/jan_verschelde/phcpy.rst:
"by perturbing the solutions"

deniederhut commented on papers/jan_verschelde/phcpy.rst:
"Parameter homotopies"

deniederhut reviewed (changes requested):
Excellent paper! I left a few inline suggestions about spelling errors and sentence structure. One general comment I have is that there is a lot of field-specific terminology here. You won't have room in this paper to explain all of them, but it could be helpful to readers if you provide descriptions for some of the more critical concepts in your library.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
"(which we gloss in this section)" is an unnecessary apology which diverts the narrative to first person. It's an eight page paper discussing a complex topic. It's inevitable that many things in the paper are glossed over. No apology is needed.

JazzTap replied:
Removed, 'per ...' left as attribution for the following discussion.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
"The number of unique mechanisms is thus a function of :math:`G` and :math:`d`, and an upper bound over :math:`d` (for which it isn't infinite) and :math:`G` with k vertices (yielding lower bounds for graphs with :math:`n \geq k` vertices) can be computed."

That is a serious run-on sentence I had to read several times to parse. Separate into at least two sentences and then parenthetical phrases will be also be unnecessary depending on whether "it" refers to the upper bound or the spatial dimension. That is, please rephrase without parenthetical phrases making use of unclear indefinite pronouns: 

"The number of unique mechanisms is thus a function of :math:`G` and :math:`d`. An upper bound over :math:`d` (for which it isn't infinite) and :math:`G` with k vertices (yielding lower bounds for graphs with :math:`n \geq k` vertices) can be computed.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
Please give a reference for the Cayley-Menger matrix.

JazzTap replied:
Cited a survey of euclidian distance geometry w/ relevant discussion (which shares a citation with Bartzos et al. of a textbook in the field, at that point).

cbcunc commented on papers/jan_verschelde/phcpy.rst:
If you are using a Latin abbreviation not in this list, https://en.wikipedia.org/wiki/List_of_Latin_abbreviations, then you should probably not be abbreviating whatever it is you are trying to say. If "s.t." is not a Latin abbreviation, it is unclear what it is and it should be explicit. I would guess "subject to," but that is ungrammatical in the sentence.

JazzTap replied:
Changed to 'such that'

cbcunc commented on papers/jan_verschelde/phcpy.rst:
"In fact" is, in fact, almost always unnecessary. Also, 

Try "by the DBSCAN class of scikit-learn."

"E.g." needs to be surrounded by commas if it is in the middle of a sentence and looks awkward. Just use ", for instance," instead.

JazzTap replied:
Applied

cbcunc commented on papers/jan_verschelde/phcpy.rst:
"et al." no period on the "et," It is a complete Latin word. Period on the "al." It is an abbreviation of _alia_.

Please split the unbalanced run-on sentence into two sentences:

Their analysis validated experimental work showing processivity in vivo. And they obtained >50x speedup over non-parameter homotopies (for running times in minutes, not hours) on systems tracking 20 paths.

JazzTap replied:
Applied throughout. Split.

cbcunc commented on papers/jan_verschelde/phcpy.rst:
Comma after "i.e."

cbcunc commented on papers/jan_verschelde/phcpy.rst:
Comma after "e.g."

cbcunc commented on papers/jan_verschelde/phcpy.rst:
Comma before "e.g."

cbcunc commented on papers/jan_verschelde/phcpy.rst:
"et al." instead of "et. al"

cbcunc commented on papers/jan_verschelde/phcpy.rst:
"one might investigate: determine" concatenantes two transitive verbs. Reword, possibly substituting the gerund form of the second verbs in the list:

determining
evaluating
describing

JazzTap replied:
Removed first transitive verb. (s/investigate/seek to/)

cbcunc commented on papers/jan_verschelde/phcpy.rst:
```diff
@@ -0,0 +1,750 @@
+
+   The solutions of a system of polynomials in several variables are often    needed, e.g.: in the design of mechanical systems, and    in phase-space analyses of nonlinear biological dynamics.    Reliable, accurate, and comprehensive numerical solutions are available    through PHCpack, a FOSS package for solving polynomial systems with    homotopy continuation.
+
+   This paper explores the development of phcpy, a scripting interface for    PHCpack, over the past five years.  One result is the availability of phcpy   through a JupyterHub featuring Python2, Python3, and SageMath kernels.
```
"through a JupyterHub **server**"

cbcunc reviewed (changes requested):
Please see the inline comments.

I ran all the code examples including the Apollonius problem (from a GItHub clone) on your JupyterHub server. The JupyterHub server with p was a big help.

Overall, this is a very good paper, if a little all over the map. I will be glad to approve it when the inline comments are addressed.

janverschelde committed:
changed the make_paper.sh back to its original state

janverschelde committed:
corrected some (not all yet) of the mistakes pointed out in the reviews,
added the example of the twisted cubic to explain the positive dimensional solution sets section better

JazzTap committed:
Revise text

janverschelde committed:
Merge pull request #4 from loldja/2019

Initial Review submission; thanks we are working on the revision.

janverschelde committed:
removed the inline high level comments

janverschelde committed:
applied the changes as in the Revise text #5

janverschelde committed:
Merge branch '2019' into 2019

janverschelde committed:
Merge pull request #5 from JazzTap/2019

Revise text,  resolved conflicts.

janverschelde committed:
Update phcpy.rst

removed extra references at the end of the paper

janverschelde committed:
removed duplicate reference SymPyDocs

janverschelde committed:
added a paragraph in the introduction describing homotopy continuation

janverschelde committed:
added references for the three classes of homotopies, mentioned free account for our JupyterHub server

janverschelde committed:
changed 'Mission' as section header into 'A Scripting Interface for PHCpack' as this is the main purpose of phcpy; some other small edits

janverschelde committed:
added thesis dissertation of Xiangcheng Yu, figured out floating placement of tables, fixed an error in the table

deniederhut commented:
@janverschelde feel free to @ us when you feel this submission is ready for re-review 😄

JazzTap committed:
Merge remote-tracking branch 'upstream/2019' into 2019

janverschelde commented:
Thanks for the detailed comments and suggestions.

We addressed most of them, I think.

The submission may indeed be ready for re-review.


________________________________
From: Dillon Niederhut <notifications@github.com>
Sent: Wednesday, June 19, 2019 6:39 PM
To: scipy-conference/scipy_proceedings
Cc: Verschelde, Jan; Mention
Subject: Re: [scipy-conference/scipy_proceedings] Paper: Solving Polynomial Systems with phcpy (#461)


@janverschelde<https://github.com/janverschelde> feel free to @ us when you feel this submission is ready for re-review 😄

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/461?email_source=notifications&email_token=AA53KBI4AMGLX4QCDQCB2FDP3K7THA5CNFSM4HN52372YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYDSNGI#issuecomment-503785113>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AA53KBNDMTLMRSNAWUUUMELP3K7THANCNFSM4HN5237Q>.

JazzTap committed:
Lint text

JazzTap commented:
I'm currently revising the abstract, besides clarifying a couple other sentences. ~~Be ready in an hour.~~ Ok.

JazzTap committed:
Add to abstract, rearrange introductions.

JazzTap committed:
Merge remote-tracking branch 'upstream/2019' into 2019

JazzTap committed:
your own -> one's own

JazzTap committed:
Update phcpy.rst

JazzTap committed:
Merge branch '2019' of https://github.com/janverschelde/scipy_proceedings into 2019

JazzTap committed:
merge references

JazzTap committed:
reword direct manipulation

JazzTap committed:
fix et al.

janverschelde committed:
Merge pull request #6 from JazzTap/2019

Finishing up responses

deniederhut commented on papers/jan_verschelde/phcpy.rst:
I still think this was meant to read "homotopies".

janverschelde committed:
fixed the hopotopies spelling mistake

deniederhut commented:
Thanks for all the edits! 

@cbcunc were there any other changes you wanted to see before approving the paper?

cbcunc reviewed (approved):
Thank you for your revisions.

deniederhut commented:
Woohoo! Thanks everyone!
